src/driver-cmstorm.c

```diff
@@ -0,0 +1,793 @@
+#include <linux/input.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <string.h>
+
+#include "libratbag-private.h"
+#include "libratbag-hidraw.h"
+#include "libratbag-data.h"
+#include "hidpp-generic.h"
+
+#define CMSTORM_NUM_BUTTONS	9
+#define CMSTORM_NUM_LEDS		3
+#define CMSTORM_NUM_PROFILES	4
+#define CMSTORM_NUM_DPI		4
+
+#define CMSTORM_REPORT_ID			0x03
+
+#define CMSTORM_REPORT_SIZE			64
+#define CMSTORM_BLOCK_SIZE			56
+#define CMSTORM_PROFILE_SIZE		0x69
+
+#define CMSTORM_CMD_SELECT_PROFILE	0xc1
+#define CMSTORM_CMD_CURRENT_PROFILE	0xc2
+#define CMSTORM_CMD_RELOAD_PROFILES	0xc5
+#define CMSTORM_CMD_WRITE_BLOCK		0xd1
+#define CMSTORM_CMD_READ_BLOCK		0xd2
+#define CMSTORM_CMD_ENABLE_PROG		0xea
+#define CMSTORM_CMD_DISABLE_PROG	0xed
+
+#define CMSTORM_BUTTON_TYPE_MOUSE		0x00
+#define CMSTORM_BUTTON_TYPE_KEYBOARD	0x02
+#define CMSTORM_BUTTON_TYPE_MACRO		0x03
+#define CMSTORM_BUTTON_TYPE_PROFILE		0x08
+#define CMSTORM_BUTTON_TYPE_DPI			0x09
+#define CMSTORM_BUTTON_TYPE_CONSUMER	0x20
+
+#define CMSTORM_MACRO_TYPE_END		0x00
+#define CMSTORM_MACRO_TYPE_MOUSE	0x01
+#define CMSTORM_MACRO_TYPE_KEYBOARD	0x03
+
+
+#pragma pack(push,1)
+struct cmstorm_button
+{
+	uint8_t repetitions;
+	uint8_t response_time;
+	uint8_t type;
+	uint16_t function;
+};
+
+struct cmstorm_dpi
+{
+	uint8_t enabled;
+	uint8_t dpi;
+	uint8_t lod;
+	uint8_t usb_rate;
+};
+
+struct cmstorm_led
+{
+	uint8_t mode;
+	uint8_t red, green, blue;
+	uint8_t intensity;
+};
+
+struct cmstorm_macro
+{
+	uint8_t type;
+	uint8_t unknown; //part of button?
+	uint8_t button;
+	uint8_t pressed_released;
+	uint8_t unknown2;
+	uint16_t delay;
+};
+
+struct cmstorm_profile
+{
+	uint8_t index;
+	uint8_t name[24]; //UTF-16BE
+	uint8_t length;
+	uint16_t macro_length;
+	struct cmstorm_button buttons[9];
+	struct cmstorm_dpi dpi[4];
+	struct cmstorm_led leds[3];
+	struct cmstorm_macro *macros;
+};
+#pragma pack(pop)
+
+static uint32_t DPIS[] = {800, 1600, 3200, 4000};
+#define DPI_COUNT 4
+
+uint32_t dpiIndexToDpi(uint8_t dpiIndex);
+uint8_t dpiToDpiIndex(uint32_t dpi);
+int send_command(struct ratbag_device *device, int cmd_length, uint8_t *cmd, int out_length, uint8_t *out);
+int reload_profiles(struct ratbag_device *device);
+int enable_profile_writing(struct ratbag_device *device);
+int disable_profile_writing(struct ratbag_device *device);
+int set_selected_profile(struct ratbag_device *device, int profile_idx);
+int read_chunk(struct ratbag_device *device, int profile, int offset, int length, uint8_t *out);
+int write_chunk(struct ratbag_device *device, int profile, int offset, int length, uint8_t *out);
+int read_chunks(struct ratbag_device *device, int profile, int offset, int length, uint8_t *out);
+int write_chunks(struct ratbag_device *device, int profile, int offset, int length, uint8_t *buf);
+int read_profile(struct ratbag_device *device, int profile, struct cmstorm_profile *out);
+int write_profile(struct ratbag_device *device, int profile, struct cmstorm_profile *out);
```

Comment:
I am not a big fan of forward declarations of every function. It's usually a useless duplication, and you can reshuffle the code so you don't need it.